File: Reference/src/Brontide.Reference.Studio/ProviderPublisherTrust.cs
using System.Buffers.Binary;
using System.Security.Cryptography;
using System.Text;

namespace Brontide.Reference.Studio;

public readonly record struct ProviderPublisherTrustPolicyId
{
    private ProviderPublisherTrustPolicyId(string value) => Value = value;

    public string Value { get; }

    public static ProviderPublisherTrustPolicyId Create(string value)
    {
        ArgumentNullException.ThrowIfNull(value);
        if (!ContentAddressedProviderStore.IsDigest(value))
        {
            throw new ArgumentException("A publisher trust policy identity must be an uppercase SHA-256 digest.", nameof(value));
        }

        return new(value);
    }
}

public enum ProviderPublisherTrustDisposition
{
    Admitted,
    Revoked,
}

public sealed record ProviderPublisherTrustEntry(
    ProviderPublisherKeyId PublisherKeyId,
    ProviderPublisherTrustDisposition Disposition);

public sealed record ProviderPublisherTrustPolicy(
    ProviderPublisherTrustPolicyId Identity,
    IReadOnlyList<ProviderPublisherTrustEntry> Entries);

public static class ProviderPublisherTrustPolicyIdentity
{
    public static ProviderPublisherTrustPolicyId Compute(IEnumerable<ProviderPublisherTrustEntry> entries)
    {
        ArgumentNullException.ThrowIfNull(entries);
        using var output = new MemoryStream();
        Append(output, "CBI35");
        var ordered = entries.OrderBy(entry => entry.PublisherKeyId.Value, StringComparer.Ordinal).ToArray();
        Append(output, ordered.Length);
        foreach (var entry in ordered)
        {
            Append(output, entry.PublisherKeyId.Value);
            Append(output, entry.Disposition switch
            {
                ProviderPublisherTrustDisposition.Admitted => "admitted",
                ProviderPublisherTrustDisposition.Revoked => "revoked",
                _ => "invalid",
            });
        }

        return ProviderPublisherTrustPolicyId.Create(Convert.ToHexString(SHA256.HashData(output.ToArray())));
    }

    private static void Append(Stream output, int value)
    {
        Span<byte> bytes = stackalloc byte[sizeof(int)];
        BinaryPrimitives.WriteInt32BigEndian(bytes, value);
        output.Write(bytes);
    }

    private static void Append(Stream output, string value)
    {
        var bytes = Encoding.UTF8.GetBytes(value);
        Append(output, bytes.Length);
        output.Write(bytes);
    }
}

public sealed record TrustedProviderPublisherAuthorization
{
    private TrustedProviderPublisherAuthorization(
        ProviderPublisherTrustPolicyId policyIdentity,
        ProviderPublisherKeyId publisherKeyId,
        ProviderArtifactSetId contentIdentity,
        string payloadSha256)
    {
        PolicyIdentity = policyIdentity;
        PublisherKeyId = publisherKeyId;
        ContentIdentity = contentIdentity;
        PayloadSha256 = payloadSha256;
    }

    public ProviderPublisherTrustPolicyId PolicyIdentity { get; }
    public ProviderPublisherKeyId PublisherKeyId { get; }
    public ProviderArtifactSetId ContentIdentity { get; }
    public string PayloadSha256 { get; }

    internal static TrustedProviderPublisherAuthorization Issue(
        ProviderPublisherTrustPolicyId policyIdentity,
        ProviderPublisherKeyId publisherKeyId,
        ProviderArtifactSetId contentIdentity,
        string payloadSha256) => new(policyIdentity, publisherKeyId, contentIdentity, payloadSha256);
}

public sealed record ProviderPublisherTrustResult
{
    private ProviderPublisherTrustResult(
        string code,
        string evidenceCode,
        ProviderPublisherTrustPolicyId policyIdentity,
        ProviderPublisherKeyId? publisherKeyId,
        ProviderArtifactSetId? contentIdentity,
        TrustedProviderPublisherAuthorization? authorization)
    {
        Code = code;
        EvidenceCode = evidenceCode;
        PolicyIdentity = policyIdentity;
        PublisherKeyId = publisherKeyId;
        ContentIdentity = contentIdentity;
        Authorization = authorization;
    }

    public string Code { get; }
    public string EvidenceCode { get; }
    public ProviderPublisherTrustPolicyId PolicyIdentity { get; }
    public ProviderPublisherKeyId? PublisherKeyId { get; }
    public ProviderArtifactSetId? ContentIdentity { get; }
    public TrustedProviderPublisherAuthorization? Authorization { get; }
    public bool IsTrusted => Authorization is not null;
    public string AdmissionCode => "admission-not-attempted";

    internal static ProviderPublisherTrustResult Refused(
        string code,
        string evidenceCode,
        ProviderPublisherTrustPolicyId policyIdentity,
        ProviderPublisherKeyId? publisherKeyId = null,
        ProviderArtifactSetId? contentIdentity = null) =>
        new(code, evidenceCode, policyIdentity, publisherKeyId, contentIdentity, null);

    internal static ProviderPublisherTrustResult Trusted(TrustedProviderPublisherAuthorization authorization) =>
        new(
            "publisher-trusted",
            "publisher-evidence-valid",
            authorization.PolicyIdentity,
            authorization.PublisherKeyId,
            authorization.ContentIdentity,
            authorization);
}

public static class ProviderPublisherTrustEvaluator
{
    internal static bool TrySnapshot(
        ProviderPublisherTrustPolicy policy,
        out ProviderPublisherTrustPolicy snapshot)
    {
        var entries = policy.Entries?.ToArray();
        if (entries is null
            || entries.Length == 0
            || entries.Any(entry => entry is null
                || string.IsNullOrWhiteSpace(entry.PublisherKeyId.Value)
                || !Enum.IsDefined(entry.Disposition))
            || entries.Select(entry => entry.PublisherKeyId).Distinct().Count() != entries.Length
            || ProviderPublisherTrustPolicyIdentity.Compute(entries) != policy.Identity)
        {
            snapshot = policy;
            return false;
        }

        snapshot = policy with { Entries = Array.AsReadOnly(entries) };
        return true;
    }

    public static ProviderPublisherTrustResult Evaluate(
        ProviderPublisherTrustPolicy policy,
        VerifiedProviderPublisherEvidence? evidence)
    {
        ArgumentNullException.ThrowIfNull(policy);
        var evidenceCode = evidence is null ? "publisher-evidence-not-verified" : "publisher-evidence-valid";
        if (!TrySnapshot(policy, out var snapshot))
        {
            return ProviderPublisherTrustResult.Refused(
                "publisher-trust-policy-invalid",
                evidenceCode,
                policy.Identity,
                evidence?.PublisherKeyId,
                evidence?.ContentIdentity);
        }

        if (evidence is null)
        {
            return ProviderPublisherTrustResult.Refused(
                "publisher-evidence-not-verified",
                evidenceCode,
                policy.Identity);
        }

        var entry = snapshot.Entries.SingleOrDefault(candidate => candidate.PublisherKeyId == evidence.PublisherKeyId);
        if (entry is null)
        {
            return ProviderPublisherTrustResult.Refused(
                "publisher-key-unknown",
                evidenceCode,
                policy.Identity,
                evidence.PublisherKeyId,
                evidence.ContentIdentity);
        }

        if (entry.Disposition == ProviderPublisherTrustDisposition.Revoked)
        {
            return ProviderPublisherTrustResult.Refused(
                "publisher-key-revoked",
                evidenceCode,
                policy.Identity,
                evidence.PublisherKeyId,
                evidence.ContentIdentity);
        }

        return ProviderPublisherTrustResult.Trusted(TrustedProviderPublisherAuthorization.Issue(
            policy.Identity,
            evidence.PublisherKeyId,
            evidence.ContentIdentity,
            evidence.PayloadSha256));
    }
}

File: Minimal/src/Brontide.Minimal.Host/ProviderPublisherTrust.fs
namespace Brontide.Minimal.Host

open System
open System.Buffers.Binary
open System.IO
open System.Security.Cryptography
open System.Text

[<StructuralEquality; StructuralComparison>]
type ProviderPublisherTrustPolicyId = private ProviderPublisherTrustPolicyId of string

[<RequireQualifiedAccess>]
module ProviderPublisherTrustPolicyId =
    let create value =
        let valid =
            not (String.IsNullOrEmpty value)
            && value.Length = 64
            && value |> Seq.forall (fun character ->
                (character >= '0' && character <= '9') || (character >= 'A' && character <= 'F'))
        if not valid then invalidArg (nameof value) "A publisher trust policy identity must be an uppercase SHA-256 digest."
        ProviderPublisherTrustPolicyId value

    let value (ProviderPublisherTrustPolicyId value) = value

type ProviderPublisherTrustDisposition =
    | Admitted
    | Revoked

type ProviderPublisherTrustEntry =
    { PublisherKeyId: ProviderPublisherKeyId
      Disposition: ProviderPublisherTrustDisposition }

type ProviderPublisherTrustPolicy =
    { Identity: ProviderPublisherTrustPolicyId
      Entries: ProviderPublisherTrustEntry list }

[<RequireQualifiedAccess>]
module ProviderPublisherTrustPolicyIdentity =
    let private appendInt32 (output: Stream) value =
        let bytes = Array.zeroCreate<byte> sizeof<int>
        BinaryPrimitives.WriteInt32BigEndian(bytes.AsSpan(), value)
        output.Write bytes

    let private appendString output (value: string) =
        let bytes = Encoding.UTF8.GetBytes value
        appendInt32 output bytes.Length
        output.Write bytes

    let compute entries =
        use output = new MemoryStream()
        appendString output "CBI35"
        appendInt32 output (List.length entries)
        entries
        |> List.sortBy (fun entry -> ProviderPublisherKeyId.value entry.PublisherKeyId)
        |> List.iter (fun entry ->
            appendString output (ProviderPublisherKeyId.value entry.PublisherKeyId)
            appendString output (match entry.Disposition with Admitted -> "admitted" | Revoked -> "revoked"))
        output.ToArray() |> SHA256.HashData |> Convert.ToHexString |> ProviderPublisherTrustPolicyId.create

type TrustedProviderPublisherAuthorization =
    { PolicyIdentity: ProviderPublisherTrustPolicyId
      PublisherKeyId: ProviderPublisherKeyId
      ContentIdentity: ProviderArtifactSetId
      PayloadSha256: string }

type ProviderPublisherTrustResult =
    { Code: string
      EvidenceCode: string
      PolicyIdentity: ProviderPublisherTrustPolicyId
      PublisherKeyId: ProviderPublisherKeyId option
      ContentIdentity: ProviderArtifactSetId option
      Authorization: TrustedProviderPublisherAuthorization option
      AdmissionCode: string }
    member this.IsTrusted = this.Authorization.IsSome

[<RequireQualifiedAccess>]
module ProviderPublisherTrustEvaluator =
    let private refused code evidenceCode (policy: ProviderPublisherTrustPolicy)
                        (keyId: ProviderPublisherKeyId option) (contentId: ProviderArtifactSetId option) =
        { Code = code
          EvidenceCode = evidenceCode
          PolicyIdentity = policy.Identity
          PublisherKeyId = keyId
          ContentIdentity = contentId
          Authorization = None
          AdmissionCode = "admission-not-attempted" }

    let evaluate (policy: ProviderPublisherTrustPolicy) (evidence: VerifiedProviderPublisherEvidence option) =
        let evidenceCode = if Option.isSome evidence then "publisher-evidence-valid" else "publisher-evidence-not-verified"
        let entries = policy.Entries |> List.ofSeq
        let structurallyValid =
            not (List.isEmpty entries)
            && (entries |> List.map _.PublisherKeyId |> List.distinct |> List.length) = entries.Length
            && ProviderPublisherTrustPolicyIdentity.compute entries = policy.Identity
        if not structurallyValid then
            refused "publisher-trust-policy-invalid" evidenceCode policy
                (evidence |> Option.map _.PublisherKeyId)
                (evidence |> Option.map _.ContentIdentity)
        else
            match evidence with
            | None -> refused "publisher-evidence-not-verified" evidenceCode policy None None
            | Some verified ->
                match entries |> List.tryFind (fun entry -> entry.PublisherKeyId = verified.PublisherKeyId) with
                | None ->
                    refused "publisher-key-unknown" evidenceCode policy
                        (Some verified.PublisherKeyId) (Some verified.ContentIdentity)
                | Some entry when entry.Disposition = Revoked ->
                    refused "publisher-key-revoked" evidenceCode policy
                        (Some verified.PublisherKeyId) (Some verified.ContentIdentity)
                | Some _ ->
                    let authorization: TrustedProviderPublisherAuthorization =
                        { PolicyIdentity = policy.Identity
                          PublisherKeyId = verified.PublisherKeyId
                          ContentIdentity = verified.ContentIdentity
                          PayloadSha256 = verified.PayloadSha256 }
                    { Code = "publisher-trusted"
                      EvidenceCode = evidenceCode
                      PolicyIdentity = policy.Identity
                      PublisherKeyId = Some verified.PublisherKeyId
                      ContentIdentity = Some verified.ContentIdentity
                      Authorization = Some authorization
                      AdmissionCode = "admission-not-attempted" }
